### PR TITLE
chore(dsl): Unifiy xroad config

### PR DESCRIPTION
**Pull request by sindrig**

### CODE REVIEW

**Review by Thorrik58 — APPROVED** [review]

**Review by arnorhs — APPROVED** [review]
that's a very welcome change

File: apps/application-system/api/infra/application-system-api.ts [hunk 1]
```diff
@@ -122,11 +104,9 @@ export const serviceSetup = (services: {
         (h) => `http://${h.svc(services.servicesPartyLetterRegistryApi)}`,
       ),
     })
+    .xroad(Base, Client, Labor, HealthInsurance, Payment)
```
Comment (saevarma) [inline apps/application-system/api/infra/application-system-api.ts:107]: `DrivingLicense` missing?

File: charts/islandis/values.staging.yaml [hunk 2]
```diff
@@ -223,16 +233,10 @@ application-system-api:
     DOCUMENT_PROVIDER_ONBOARDING_REVIEWER: '/k8s/application-system/api/DOCUMENT_PROVIDER_ONBOARDING_REVIEWER'
     DOKOBIT_ACCESS_TOKEN: '/k8s/application-system/api/DOKOBIT_ACCESS_TOKEN'
     DOKOBIT_URL: '/k8s/application-system-api/DOKOBIT_URL'
-    DRIVING_LICENSE_SECRET: '/k8s/application-system/api/DRIVING_LICENSE_SECRET'
```
Comment (saevarma) [inline charts/islandis/values.staging.yaml:226]: Isn't `XROAD_DRIVING_LICENSE` missing? Same for prod and dev files. Added another comment on the corresponding infra file for `application-system-api`

**Review by saevarma — CHANGES_REQUESTED** [review]
Great work! 
Just two minor comments and one question if `DrivingLicense` got forgotten for `application-system-api`

File: infra/src/dsl/dsl.ts [hunk 3]
```diff
@@ -96,17 +98,35 @@ export class ServiceBuilder<ServiceType> implements Service {
...
    * @param key name of env variable
    * @param value value of env variable. A single string sets the same value across all environment. A dictionary with keys the environments sets an individual value for each one
    */
   env(envs: EnvironmentVariables) {
+    this.assertUnset(this.serviceDef.env, envs)
     this.serviceDef.env = { ...this.serviceDef.env, ...envs }
     return this
   }
 
+  xroad(...configs: XroadConfig[]) {
```
Comment (eirikurn) [inline infra/src/dsl/dsl.ts:125]: Missing JSDoc?

Is the DSL documented somewhere in the handbook or just JSDoc?

**Review by eirikurn — APPROVED** [review]
Only reviewed the concept. I trust that you/others verify that there are no regressions with environment variables.

My main comment/suggestion would be to expand this to cover more "shared config" use cases, or just more "integrations".

```ts
// or .configs([
.integrations([
  XroadBase,
  XroadClient,
  DrivingLicense,
  Contentful,
  Etc,
])
```

This would nicely complement the ConfigModule stuff I'm speccing here: https://github.com/island-is/internal/issues/19

File: infra/src/dsl/xroad.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
+import { EnvironmentVariables, Secrets } from './types/input-types'
+
+type XroadSectionConfig = {
+  secrets?: Secrets
+  env?: EnvironmentVariables
+}
+
+export class XroadConf {
```
[the remark this answers is not in the document]
Comment (sindrig) [reply infra/src/dsl/xroad.ts:8]: yep!

File: infra/src/dsl/types/input-types.ts [hunk 5]
```diff
@@ -28,7 +28,22 @@ export type HealthProbe = {
 }
 
 export type Secrets = { [name: string]: string }
+export type Environment = { [name: string]: string }
```
[the remark this answers is not in the document]
Comment (sindrig) [reply infra/src/dsl/types/input-types.ts:31]: yep, using EnvironmentVariables instead

File: charts/islandis/values.staging.yaml (hunk 2 above)
Comment (sindrig) [reply charts/islandis/values.staging.yaml:226]: whew nice catch. added. thanks!

File: infra/src/dsl/dsl.ts (hunk 3 above)
Comment (sindrig) [reply infra/src/dsl/dsl.ts:125]: added jsdoc. @pshomov is there any dsl documentation in the handbook?

**Review by saevarma — APPROVED** [review]
LGTM

**Comment by eirikurn** [comment]
Oh and also.

https://media.giphy.com/media/eM0U5NQtVHu30VM5sS/giphy.gif

🎉🎉🎉

**Review by pshomov — APPROVED** [review]
LGTM

File: infra/src/dsl/dsl.ts (hunk 3 above)
Comment (pshomov) [reply infra/src/dsl/dsl.ts:129]: Have to check, it is mentioned in the Helm docs where we are discussing the configuration of services

**Review by gudjong — APPROVED** [review]
Looks good.

**Review by hinriksteinar — APPROVED** [review]
LGTM!

**Pull request merged**